Add support for Nicira Openflow extension

wenyingd opened the pull request:
1. Add group support. Group is an implementation of FgraphElem, and support to add buckets
2. Support NX extension of actions: resubmit, output to NX register. These two actions are
   working as implementation of FgraphElem
3. Add EmptyElem to support none following output actions, used for multiple conjunction case
4. Add support for Nicira extended actions: load, move, conjunction, ct
5. Add support for Nicira extented match: regX[m.n], ct_state, ct_mark
6. This change is denpendant on change of libOpenflow to add support for Nicira extensions

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Add support to group entry

1. Group is working as one fgrapElem, it could be a `Next` for any Flow
entry.
2. Group supports to generate an action for flow entry, and send GroupMod
messages to OFSwitch.
3. Group object only support to add buckets or reset buckets. Client could
maintain all valid buckets and reset if it wants to remove some specific
bucket.

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Enhance `output` action

1. Support to output to in_port, which could be used in some responder,
   e.g., ARP Responder
2. Wrap a function provide action "normal", so that consumer could not
   know message definition for 'normal' action
3. Add support to output an OVS register field, which could be invoked
   if using register to cache ofport
4. Add support for OVS `resubmit` actions, which should also be a
   fgraphElem, and it could work as `Next` for flow entry

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Enhance flow entry match and actions

1. Add support to match with arp,conj_id,ct_mark,ct_state,regX[m..n]
2. Add support to set actions: load, move ct, conjunction
3. Leave flow cookie id configurable by client. Using global flow ID
   if client not set it.
4. Add an empty fgraphElem for flow entries not use output nor group
   as the last actions, e.g., conjunction

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Enhance flow to support monitor flow entry realize status

Leverage MultipartMessage to retrieve status of new added flow, set realize
status as true once got reply for this request, and then the request id
should be removed from switch cache.

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Enhance controller to support working with Unix Domain Socket

Controller could leverage Unix Domain Socket to connect and control local
ofSwitch. In this case, controller should work as connection initiator, and
it manages only one switch. Once connnection on unix domain socket is
disconnected, controller should retry the connection

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Provide public functions to reset Flow.isInstalled value

Signed-off-by: wenyingd <wenyingd@vmware.com>

blp commented on ofctrl/fgraphResubmit.go:
```diff
@@ -0,0 +1,41 @@
+
+// Return a resubmit action (Used as a last action by flows in the table pipeline)
+func (self *Resubmit) GetResubmitAction() openflow13.Action {
+	if self.ofport == 0 {
```
I don't think that 0 is a good choice for the dummy value, because it's occasionally used as a real resubmit port. Is there a way to use some other value?

wenyingd replied:
I would move this check to the construct func, and change the func `NewResubmit(inPort uint16, table uint8)` as `NewResubmit(inPort *uint16, table *uint8)` . Then the default port or table number would be set when the parameter is nil.

blp commented on ofctrl/fgraphResubmit.go:
```diff
@@ -0,0 +1,41 @@
+	if self.ofport == 0 {
+		self.ofport = openflow13.OFPP_IN_PORT
+	}
+	if self.nextTable == 0 {
```
0 is definitely not a good choice for a dummy value because resubmit to table 0 is common.

wenyingd replied:
Thanks for catching this issue, I would change to check if the input number is nil in the construction, and leverage  OFPP_IN_PORT and OFPTT_ALL to set corresponding field.

blp commented on ofctrl/fgraphNXOutput.go:
```diff
@@ -0,0 +1,43 @@
+// This file implements the forwarding graph API for output to NX register element
+
+type NXOutput struct {
+	field      *openflow13.MatchField // Target OXM/OXX field
```
I think you mean OXM/NXM

wenyingd replied:
Yes, I will change.

blp commented on ofctrl/fgraphNXOutput.go:
```diff
@@ -0,0 +1,43 @@
+
+type NXOutput struct {
+	field      *openflow13.MatchField // Target OXM/OXX field
+	fieldRange *openflow13.NXRange    // Field range of target register to resubmit
```
"output", not "resubmit"

wenyingd replied:
Yes, I will change.

blp commented on ofctrl/fgraphGroup.go:
```diff
@@ -0,0 +1,107 @@
+func (self *Group) Delete() error {
+	groupMod := openflow13.NewGroupMod()
+	groupMod.GroupId = self.ID
+
```
I suspect these can be moved inside the if statement since the groupmod is only used if the group is installed.

wenyingd replied:
I will update.

blp commented on ofctrl/fgraphGroup.go:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	// Mark it as unInstalled
+	self.isInstalled = false
```
It seems reasonable to move this inside the 'if' also.

wenyingd replied:
I will update.

blp commented on ofctrl/ofSwitch.go:
```diff
@@ -304,3 +326,28 @@ func (self *OFSwitch) DumpFlowStats(cookieID, cookieMask uint64, flowMatch *Flow
+				return
+			} else {
+				time.Sleep(50 * time.Millisecond)
+			}
```
I don't really understand what's going on here, but it looks like a loop with a sleep in it that's waiting for a reply across a Unix domain socket. This is odd. I would normally use select() or poll() to wait for a reply, or a blocking read.

wenyingd replied:
I would add a channel to notify the controller has received echoReply message, and lever select() to wait for the notification or timeout event.

blp commented on ofctrl/ofctrl.go:
```diff
@@ -99,15 +126,80 @@ func (c *Controller) Listen(port string) {
+
+			conn, err = net.Dial("unix", sock)
+			if err != nil {
+				log.Fatal(err)
```
Is it really a good idea to kill the controller if the connection fails?

wenyingd replied:
I would add retry here. However, if the connection still fails after max retry, I think we should kill the controller as it is not available. What do you think?

blp commented:
Also: this appears to be a pull request from your repository to your repository. Did you mean for it to be a pull request to the upstream contiv/ofnet repository?

wenyingd commented:
We used to think it would be better to not expose internal comments to the ofnet community. So I closed that PR, and planned to create a new one after the internal review is done.

wenyingd committed:
Revise method `OFSwitch.CheckStatus`

Add a channel to notify EchoReply message is received after sending EchoRequest
messagestatus to check connection status. The notification is broadcasted by
closing the channel, and then set the channel as nil. When the client checks
the connection status, it would try to re-create the channel.

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Reorder FlowActions when construct FlowMod message

Prepend FlowAction into the instruction in the reverse order, so that the actions are
in the order as they are added into Flow, and the action in the NextElem is the last
action of the instruction.

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Bump up libOpenflow version
